# Probability: Introduction to Foundational Concepts


## Objectives

- Describe set theory and its terminology
- Recognize combinatorics
- Describe the fundamentals of probability theory
- Define conditional probability


![xkcd](images/increased_risk_2x.png)
[[Image Source]](https://xkcd.com/1252/)

# Prelude:

## Set Theory Basics

In probability theory, a **set** is a well-defined collection of objects (called **elements**). An element is either in a given set or not, and order is not considered.

> If an **element** $x$ belongs to a set $S$, then you'd write $x \in S$. 
> 
> On the other hand, if $x$ does not belong to $S$, then you'd write $x\notin S$.

Although sets can be abstract mathematical object, we'll typically think of sets as a collection of things sharing some relevant attribute(s).

### Set Notation

<img src="images/new_venn_diagram.png" alt="venn diagram with set notation" width=500>

The **union** of 2 sets $A$ and $B$ is the set $T$ of elements in either $A$ or $B$, or in both. We denote this with the symbol $\cup$.


$$\large T = A\cup B = B\cup A$$

The **intersection** of two sets $A$ and $B$ is the set $V$ that contains all elements of $A$ that also belong to $B$. We denote this with the symbol $\cap$.

$$\large V = A\cap B = B\cap A$$

<img src="images/setnotation.jpg" alt="set notation, found from https://slideplayer.com/slide/10502152/" width=800>

[Image Source](https://slideplayer.com/slide/10502152/)

More on Sets: https://www.mathsisfun.com/sets/sets-introduction.html

### But Now, in Python:

 These are set methods - work only on Python sets!

| Method                      | Result |
| ------                      | ------ |
| `s.issubset(t)`             | test whether every element in s is in t |
| `s.issuperset(t)`           | test whether every element in t is in s |
| `s.union(t)`                | new set with elements from both s and t |
| `s.intersection(t)`         | new set with elements common to s and t |
| `s.difference(t)`           | new set with elements in s but not in t |
| `s.symmetric_difference(t)` | new set with elements in either s or t but not both |

## Combinatorics

Before we dive in much further, let's briefly talk about how you can take elements from different collections and group them.

### Combinatorics Definition

> Combinatorics is the branch of mathematics that deals with the relations characterizing sets, subsets, lists, and multisets.
>
> Sometimes combinatorics is said to be the branch of math that deals with counting; and that’s true, but not in the sense in which you learned to count in kindergarten. Though combinatorics deals with numbering and finding out how many members are in sets, it’s designed to find ways of doing that **without actual, potentially tedious, counting** involved.

-- [Statistics How To](https://www.statisticshowto.com/probability-and-statistics/combinatorics/)

### What's the difference between a permutation and a combination?

In a **permutation**, order matters. If you have a race, it matters who arrives in first, or second, or third place - there's a difference in the ordering of the group!

> The number of **permutations** of **n** objects taken **r** at a time is given by the formula:
>
> $$\large P(n,r) = \frac{n!}{(n - r)!}$$

In a **combination**, you only care about which items are members of the set. For example, if you're creating groups of students to work on a project, the order in which you add their names to the group doesn't really matter - it's the group members itself, not any order, that you care about.

> The number of **combinations** of **n** objects taken **r** at a time is given by the formula:
>
> $$\large C(n,r) = \frac{n!}{r!(n - r)!}$$

Main things to ask when dealing with combinations or permutations:
- Does order matter? 
- With or without replacement? (aka does the first choice then restrict the second choice?)

### Q: How many possible orders are there for first/second/third place in a race with 30 contestants?

In [1]:
# Can use math:
import math

n = 30
r = 3

math.factorial(n) / math.factorial(n-r)

24360.0

In [2]:
# Also the python library itertools has combination and permutations!
import itertools

len(list(itertools.permutations(range(1,31), 3)))

24360

### Q: How many possible codes are there for a standard padlock?

> Hint: (there are 40 numbers on a padlock and use 3 numbers - **with replacement**)

In [10]:
# A:
40**3



# Why can't we use itertools?
# Because their permutation function doesn't allow for replacement (alas)

64000

Formula here is:

$$ n ^ r $$

For the first number: 40 choices. For the second number, still 40 choices: $40\cdot40=40^2$. Again, for 3rd number, still 40 choices: $40\cdot40\cdot40=40^3$

### Q: How many unique 3 topping pizzas can you make from the following 8 ingredients:

- Mushrooms
- Pepperoni
- Onion
- Peppers
- Ham
- Pineapple
- Sausage
- Olives
    
> Side note: which is the worst?

In [11]:
# Using math

math.factorial(8) / (math.factorial(3) * math.factorial(8-3))

56.0

In [12]:
# Our list of toppings
toppings = ["Mushrooms", "Pepperoni", "Onion", "Peppers", 
            "Ham", "Pineapple", "Sausage", "Olives"]

# Using itertools
three_topping_pizzas = list(itertools.combinations(toppings, 3))
len(three_topping_pizzas)

56

In [13]:
three_topping_pizzas

[('Mushrooms', 'Pepperoni', 'Onion'),
 ('Mushrooms', 'Pepperoni', 'Peppers'),
 ('Mushrooms', 'Pepperoni', 'Ham'),
 ('Mushrooms', 'Pepperoni', 'Pineapple'),
 ('Mushrooms', 'Pepperoni', 'Sausage'),
 ('Mushrooms', 'Pepperoni', 'Olives'),
 ('Mushrooms', 'Onion', 'Peppers'),
 ('Mushrooms', 'Onion', 'Ham'),
 ('Mushrooms', 'Onion', 'Pineapple'),
 ('Mushrooms', 'Onion', 'Sausage'),
 ('Mushrooms', 'Onion', 'Olives'),
 ('Mushrooms', 'Peppers', 'Ham'),
 ('Mushrooms', 'Peppers', 'Pineapple'),
 ('Mushrooms', 'Peppers', 'Sausage'),
 ('Mushrooms', 'Peppers', 'Olives'),
 ('Mushrooms', 'Ham', 'Pineapple'),
 ('Mushrooms', 'Ham', 'Sausage'),
 ('Mushrooms', 'Ham', 'Olives'),
 ('Mushrooms', 'Pineapple', 'Sausage'),
 ('Mushrooms', 'Pineapple', 'Olives'),
 ('Mushrooms', 'Sausage', 'Olives'),
 ('Pepperoni', 'Onion', 'Peppers'),
 ('Pepperoni', 'Onion', 'Ham'),
 ('Pepperoni', 'Onion', 'Pineapple'),
 ('Pepperoni', 'Onion', 'Sausage'),
 ('Pepperoni', 'Onion', 'Olives'),
 ('Pepperoni', 'Peppers', 'Ham'),
 ('Pepper

# Probability Theory

Probability theory is the study of the frequency of a given event occurring with respect to all possible events.

**Probability is the likelihood of a specific outcome occurring out of all possible outcomes, expressed as a fraction between 0 and 1.**

But perhaps more importantly:

> **"Probabilities do not tell us what will happen for sure; they tell us what is _likely to happen_ and what is _less likely to happen_."**
>
> -- _Naked Statistics_, by Charles Wheelan, p. 72

## Why Should I Care About Probabilities?

Studying probabilities allows us to make better and more informed decisions, based on data previously collected. For example, understanding the fact that it is nearly impossible for us to ever win the lottery from a probabilistic standpoint deters us from using it as a source of income.

Probability theory also lies at the heart of making inferences using our data, which is what statistics is all about!

## Probability Terminology

From [_Introduction to Probability, Statistics, and Random Processes_](https://www.probabilitycourse.com/chapter1/1_3_1_random_experiments.php):

> **Outcome:** A result of a random experiment.
>
> **Sample Space:** The set of all possible outcomes.
>
> **Event:** A subset of the sample space.

To be more specific, an **event** is a specific outcome of a random experiment, often what we are hoping to predict (so it's sometimes called a successful outcome)

We typically write probability statements using notation like $P(E) = 0.5$, where $E$ is an event, such as getting Heads in a coin flip.

## Probability Using Sets

In its most basic form, the **probability** of an event occurring is the size of the specific outcome set divided by the size of the sample space. In some ways, probability is just about counting outcomes.

**Example**: What is the probability of rolling an even number on a six-sided die?

* Event: {2, 4, 6} has size 3
* Sample space: {1, 2, 3, 4, 5, 6} has size 6
* Thus the probability = 3/6 = 1/2 = 50%

#### Relevant Distinction: Types of Variables

A random variable can either be continuous or discrete
- **Continuous**: the variable can have any values within a range (e.g. a person's height)
- **Discrete**: the variable can only have a finite subset of values within a range (e.g. number of cars owned)

This is important because, for continuous variables, the probability of some specific event x _will functionally be zero_! BUT "in reality we are always interested in the probability of some intervals rather than a specific point x" [[Source]](https://www.probabilitycourse.com/chapter1/1_3_5_continuous_models.php)

## Calculating Probabilities

Note: some of this text comes from [_OpenIntro Statistics_](https://www.openintro.org/book/os/), Chapter 3.

In general, you can think of dividing the outcome you're exploring by all possible outcomes:

$$\large P(Event) = \frac{|Event|}{|Sample\ Space|} $$

#### General Addition Rule

The probability that either $A$ or $B$ will occur can be calculated by adding each individual probability, and then subtracting the probability that both occur together:

$$\large P(A \cup B) = P(A) + P(B) − P(A \cap B)$$

Remember, $P(A \cap B)$ expresses the overlap between the two events - if you don't subtract that overlap, then you double count the instances when **both** $A$ and $B$ occur!

#### Multiplication Rule for Independent Processes

A special condition is when the outcome of $A$ has no bearing on the outcome of $B$. We say these two events are **independent** (e.g. rolling a die and tossing a coin).

If $A$ and $B$ represent events from two different and **independent** processes, then the probability
that both $A$ and $B$ occur can be calculated as the product of their separate probabilities:

$$\large P(A \cap B) = P(A) * P(B)$$

### 🧠 Knowledge Check

Die Options: 1 - 2 - 3 - 4 - 5 - 6

Coin Options: Heads - Tails

Total Options:

- 1 & Heads
- 2 & Heads
- 3 & Heads
- 4 & Heads
- **5** & Heads
- 6 & Heads

- 1 & **Tails**
- 2 & **Tails**
- 3 & **Tails**
- 4 & **Tails**
- **5** & **Tails**
- 6 & **Tails**

### 1) AND Question:

What is the probability of rolling a 5 on a fair die _and_ getting a tails on a fair coin toss?

**Your answer here**:

- P(5) = 1/6
- P(tails) = 1/2

P( 5 & tails) = 1/6 * 1/2 = 1/12


In [20]:
(1/6) * (1/2) == 1/12

True

### 2) OR Question: 

What is the probability of rolling a 5 on a die _or_ getting a tails on a coin toss?

**Your answer here**:

- P(5) = 1/6
- P(tails) = 1/2

P(5 OR tails) = 1/6 + 1/2 - 1/12
              = 2/12 + 6/12 - 1/12 = 7/12


In [23]:
(1/6) + (1/2) - (1/12)

0.5833333333333333

In [24]:
7/12

0.5833333333333334

## Enough Talk - Let's Explore in Python!

### Mushroom dataset

Let's look at a modified version of the Mushroom dataset from UCI [here](https://archive.ics.uci.edu/ml/datasets/Mushroom). Each row in this dataset corresponds to one observation (one mushroom). 

In [25]:
import pandas as pd

In [26]:
df = pd.read_csv('data/Mushrooms_cleaned.csv')
df.head()

edible-poisonous gill-spacing stalk-shape stalk-color-above-ring  \
0        poisonous        close   enlarging                  white   
1           edible        close   enlarging                  white   
2           edible        close   enlarging                  white   
3        poisonous        close   enlarging                  white   
4           edible      crowded    tapering                  white   

  stalk-color-below-ring gill-color  bruised  
0                  white      black     True  
1                  white      black     True  
2                  white      brown     True  
3                  white      brown     True  
4                  white      black    False

In [27]:
df.describe()

edible-poisonous gill-spacing stalk-shape stalk-color-above-ring  \
count              8124         8124        8124                   8124   
unique                2            2           2                      9   
top              edible        close    tapering                  white   
freq               4208         6812        4608                   4464   

       stalk-color-below-ring gill-color bruised  
count                    8124       8124    8124  
unique                      9         12       2  
top                     white       buff   False  
freq                     4384       1728    4748

#### 1) If you picked a row from this dataset at random, what is the probability it corresponds to a bruised mushroom? 

In other words, find $P(bruised)$

In [29]:
df['bruised'].sum()

3376

In [30]:
p_bruised = df['bruised'].sum() / len(df)

In [31]:
p_bruised

0.4155588380108321

In [36]:
bruised = df.loc[df['bruised'] == True]

In [35]:
len(bruised) / len(df)

0.4155588380108321

In [37]:
len(df)

8124

In [38]:
# Let's see...
df.sample(1)

edible-poisonous gill-spacing stalk-shape stalk-color-above-ring  \
5132        poisonous        close    tapering                  white   

     stalk-color-below-ring gill-color  bruised  
5132                   pink       buff    False

#### 2) What is the probability you pick a row corresponding to a mushroom that is bruised _AND_ edible?

$P(edible \cap bruised)$

In [41]:
df.head(3)

edible-poisonous gill-spacing stalk-shape stalk-color-above-ring  \
0        poisonous        close   enlarging                  white   
1           edible        close   enlarging                  white   
2           edible        close   enlarging                  white   

  stalk-color-below-ring gill-color  bruised  
0                  white      black     True  
1                  white      black     True  
2                  white      brown     True

In [46]:
brusied_AND_edible = df.loc[(
    df['edible-poisonous'] == 'edible') & (df['bruised'] == True)]

In [48]:
p_brusied_AND_edible = len(brusied_AND_edible) / len(df)

p_brusied_AND_edible

0.33874938453963566

BUT! Are they independent events ...?

> Formally, $A$ and $B$ are *independent* if and only if the probability that *both* $A$ *and* $B$ happen is:
> 
> $$\large P(A \cap B) = P(A) * P(B)$$

In [53]:
p_bruised

0.4155588380108321

In [54]:
p_edible = (len(df.loc[df['edible-poisonous'] == 'edible']))/len(df)
p_edible

0.517971442639094

In [55]:
p_bruised * p_edible

0.21524761082589627

In [58]:
p_brusied_AND_edible

0.33874938453963566

In [57]:
p_brusied_AND_edible == p_bruised * p_edible

False

## Enter: Conditional Probability

### When do we compute conditional probabilities? 

We need to compute conditional probabilities when the outcome of an event depends on the outcome of previous events (**dependent** events). A conditional probability of an event is the probability of the event *given* another event has occurred.


When events _are_ independent, the rule for probabilistic AND (I'll use '$\cap$' below) is simple:

$$\large P(A\cap B) = P(A) * P(B)$$

But the more general rule, which includes non-independent events, is:

$$\large P(A\cap B) = P(A | B) * P(B)$$

> (this is pretty much the law of total probability, we'll revisit this later!)

In fact, this is the definition of conditional probability. Rearranging:

$$\large P(A | B) = \frac{P(A\cap B)}{P(B)}$$

The `|` here should be read as "given". We are given some information, $B$, and thus it reduces our sample space!

### Back to our Mushrooms!

In [59]:
p_brusied_AND_edible
# intersection!

0.33874938453963566

In [63]:
df.loc[(df['bruised'] == True) & (df['edible-poisonous'] == 'edible')]

edible-poisonous gill-spacing stalk-shape stalk-color-above-ring  \
1              edible        close   enlarging                  white   
2              edible        close   enlarging                  white   
5              edible        close   enlarging                  white   
6              edible        close   enlarging                  white   
7              edible        close   enlarging                  white   
...               ...          ...         ...                    ...   
7940           edible        close   enlarging                  white   
7946           edible        close   enlarging                  white   
7952           edible        close   enlarging                  white   
7965           edible        close   enlarging                  white   
8038           edible        close   enlarging                  white   

     stalk-color-below-ring gill-color  bruised  
1                     white      black     True  
2                     white      brown     True  
5                     white      brown     True  
6                     white       gray     True  
7                     white      brown     True  
...                     ...        ...      ...  
7940                  white      white     True  
7946                  white      white     True  
7952                  white      white     True  
7965                  white      white     True  
8038                  white      white     True  

[2752 rows x 7 columns]

In [62]:
len(df.loc[(df['bruised'] == True) & (df['edible-poisonous'] == 'edible')]) / len(df)

0.33874938453963566

#### 3) What is the probability of picking an edible mushroom given it is bruised? 

$P(edible | bruised)$

In [65]:
p_edible_given_bruised = p_brusied_AND_edible / p_bruised
p_edible_given_bruised

0.8151658767772512

In [69]:
len(bruised)

3376

In [70]:
len(bruised.loc[bruised['edible-poisonous'] == 'edible'])

2752

In [71]:
len(bruised.loc[bruised['edible-poisonous'] == 'edible'])/len(bruised)

0.8151658767772512

#### 4) What is the probability of picking a bruised mushroom given it is edible? 

$P(bruised | edible)$

In [72]:
p_bruised_given_edible = p_brusied_AND_edible/p_edible
p_bruised_given_edible

0.6539923954372624

### Intuition behind conditional probability: 

How do you compute the probability that mushrooms are edible given they are bruised? 

When you ask the question "what is the probability that the mushrooms are edible and bruised?", the sample space originally contains all 8124 rows of mushrooms. 

<img src="images/Image_72_Cond4.png" width="300">

However, to compute the probability that the mushrooms are edible given they are bruised, you need to consider the reduced size of the sample space. 

In the image above, S is the universe of all mushrooms in the dataset, A is the set of mushrooms that are edible, and B is the set of mushrooms that are bruised.

* When you ask the question "what is the probability that the mushrooms are edible given the mushrooms are bruised?", you have effectively reduced the size of the sample space to include only those mushrooms that are bruised. 

* Given that mushrooms are bruised, the only way for the mushrooms to be edible is for these mushrooms to fall in the intersection of the set of mushrooms that are edible _and_ the set of mushrooms that are bruised , $P(edible \cap bruised)$.  

* To account for the smaller sample space, you divide the probability mushrooms are edible and bruised by the probability the mushrooms are bruised: $$\large P(edible|bruised) = \frac{P(edible \cap bruised)}{P(bruised)}$$




## Law of Total Probability

Sometimes we want to calculate an **unconditional** probability by making use of some **conditional** probabilities. Enter the Law of Total Probability:

$$\large P(A)=\sum _{n}P(A\cap B_{n})$$

Which works out to: 

$$\large P(A)=\sum _{n}P(A\mid B_{n}) * P(B_{n})$$

Where $B_{n}$ captures all of the disjoint events that combine to define the sample space, so $\sum P(B_{n}) = 1$.

### In practice:

Let's say we want to know the likelihood that a mushroom is **edible**. If we know the likelihood that all tapering stalk mushrooms are edible, and the likelihood that not-tapering-stalk mushrooms are edible, plus how likely it is that mushrooms have tapering stalks (and thus also how likely the mushroom does _not_ have a tapering stalk, the compliment of the mushrooms having a tapering stalk) - well, we can use all of that to find the likelihood our mushroom will be edible! 

$$\large P(\text{edible}) = P(\text{edible} | \text{tapering}) * P(\text{tapering}) + P(\text{edible} | \text{not tapering}) * P(\text{not tapering})$$

We can do this because 'tapering stalk' and 'not tapering stalk' capture all of the possibilities of mushrooms - they define our entire sample space!

In [73]:
df.head()

edible-poisonous gill-spacing stalk-shape stalk-color-above-ring  \
0        poisonous        close   enlarging                  white   
1           edible        close   enlarging                  white   
2           edible        close   enlarging                  white   
3        poisonous        close   enlarging                  white   
4           edible      crowded    tapering                  white   

  stalk-color-below-ring gill-color  bruised  
0                  white      black     True  
1                  white      black     True  
2                  white      brown     True  
3                  white      brown     True  
4                  white      black    False

In [74]:
df.loc[df['stalk-shape'] == 'tapering']

edible-poisonous gill-spacing stalk-shape stalk-color-above-ring  \
4              edible      crowded    tapering                  white   
14             edible      crowded    tapering                  white   
16             edible      crowded    tapering                  white   
29             edible      crowded    tapering                  white   
35             edible      crowded    tapering                  white   
...               ...          ...         ...                    ...   
8113        poisonous        close    tapering                   pink   
8116        poisonous        close    tapering                   pink   
8117        poisonous        close    tapering                   pink   
8118        poisonous        close    tapering                   pink   
8122        poisonous        close    tapering                  white   

     stalk-color-below-ring gill-color  bruised  
4                     white      black    False  
14                    white      brown    False  
16                    white      black    False  
29                    white      brown     True  
35                    white      white     True  
...                     ...        ...      ...  
8113                   pink       buff    False  
8116                  white       buff    False  
8117                  white       buff    False  
8118                  white       buff    False  
8122                  white       buff    False  

[4608 rows x 7 columns]

In [75]:
# First let's find our tapering and not-tapering subsets
tapering = df.loc[df['stalk-shape'] == 'tapering']

not_tapering = df.loc[df['stalk-shape'] != 'tapering']

In [78]:
# Calculate our probability of tapering, and probability of not-tapering
p_tapering = len(tapering)/len(df)

p_not_tapering = len(not_tapering)/len(df)
# same as 
# p_not_tapering = 1 - p_tapering

In [79]:
# Now calculate our two conditional probabilties
p_edible_given_tapering = len(
    tapering.loc[tapering['edible-poisonous'] == 'edible'])/len(tapering)

p_edible_given_not_tapering = len(
    not_tapering.loc[not_tapering['edible-poisonous'] == 'edible'])/len(not_tapering)

In [80]:
# Do the math!
(p_edible_given_tapering * p_tapering) + (p_edible_given_not_tapering * p_not_tapering)

0.517971442639094

In [81]:
# Sanity check
p_edible

0.517971442639094

## Partitioning Complex Events

You're not really a mushroom expert, but you can see that the mushroom in your hand is has some orange on the stalk. Given the data at your disposal, what's the probability that the mushroom is edible?



$$\large P(\text{edible|orange stalk}) = \frac{P(\text{edible} \cap \text{orange stalk})}{P(\text{orange stalk})}$$

Furthermore, we can decompose $P(\text{orange stalk})$ into the two possibilities:

$P(\text{orange stalk}) = P(\text{orange stalk below ring}\cup\text{orange stalk above ring})$

But be careful here! 

$P(\text{orange stalk below ring}\cup \text{orange stalk above ring})$ **DOES NOT EQUAL** $P(\text{orange stalk below ring}) + P(\text{orange stalk above ring})$

While this may seem correct, adding these individual probabilities **double counts** mushrooms which have entirely orange stalks. However - when we do these things in Pandas (using a loc and an or condition), it won't duplicate the row - so you don't need to worry about it here!

In [82]:
# Can prove that P(orange above) + P(orange below) != p(orange stalk)
p_orange_above = df[df['stalk-color-above-ring'] == 'orange'].shape[0]/df.shape[0]
p_orange_below = df[df['stalk-color-below-ring'] == 'orange'].shape[0]/df.shape[0]
# .shape[0] is the same as calculating len()

In [85]:
p_orange_above + p_orange_below

0.047267355982274745

In [86]:
p_orange_stalk = df[(df['stalk-color-above-ring'] == 'orange')
                    | (df['stalk-color-below-ring'] == 'orange')
                    ].shape[0]/df.shape[0]
p_orange_stalk

0.023633677991137372

In [84]:
p_orange_stalk == (p_orange_above + p_orange_below)

False

In [ ]:
# Now let's find P(edible | orange stalk)


# Apparently orange-stalk mushrooms seem fairly safe....
# (Disclaimer: don't take this as definitive foraging adivce!)

In [ ]:
# Sanity check...


## Extra Credit: Additional Conditional Probability Practice

What's the probability that a mushroom is poisonous if it has close gill spacing AND a tapering stalk?

$$\large P(poisonous|close \cap tapering) = \frac{P(poisonous \cap close \cap tapering)}{P(close \cap tapering)}$$

In [ ]:
# P that mushroom is poisonous given close gill spacing and tapering stalk
